tests/test_selector.py

```diff
@@ -1176,6 +1180,73 @@ def test_deep_nesting(self):
         self.assertEqual(len(sel.css("span")), nest_level)
         self.assertEqual(len(sel.css("td")), 1)
 
+    def test_invalid_type(self):
+        with self.assertRaises(ValueError):
+            self.sscls("", type="xhtml")
+
+    def test_default_type(self):
+        text = "foo"
+        selector = self.sscls(text)
+        self.assertEqual(selector.type, "html")
+
+    def test_json_type(self):
+        obj = 1
+        selector = self.sscls(str(obj), type="json")
+        self.assertEqual(selector.root, obj)
+        self.assertEqual(selector.type, "json")
+
+    def test_html_root(self):
+        root = etree.fromstring("<html/>")
+        selector = self.sscls(root=root)
+        self.assertEqual(selector.root, root)
+        self.assertEqual(selector.type, "html")
+
+    def test_json_root(self):
+        obj = 1
+        selector = self.sscls(root=obj)
+        self.assertEqual(selector.root, obj)
+        self.assertEqual(selector.type, "json")
+
+    def test_json_xpath(self):
+        obj = 1
+        selector = self.sscls(root=obj)
+        with self.assertRaises(ValueError):
+            selector.xpath("//*")
+
+    def test_json_css(self):
+        obj = 1
+        selector = self.sscls(root=obj)
+        with self.assertRaises(ValueError):
+            selector.css("*")
+
+    def test_invalid_json(self):
+        text = "<html/>"
+        selector = self.sscls(text, type="json")
+        self.assertEqual(selector.root, None)
```

Comment:
Based on tests, it seems that when you pass `type="json"`, selector.root can be either the loaded object (`1` in tests above), or a string to parse using JSON parser. 

Sometimes a loaded object could be a string, right? Is it intentional that we require it to be json-parsable to be set as root?

Reply:
If you want root to be a string value, you must use either `Selector(root="foo")` or `Selector('"foo"')`. Both should work. I have added a test for all JSON types now, that checks both when they are passed as text or as root.